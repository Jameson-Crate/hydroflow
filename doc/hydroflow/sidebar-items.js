window.SIDEBAR_ITEMS = {"attr":["main","test"],"derive":["DemuxEnum"],"enum":["Never"],"macro":["assert_var_impl","datalog","hydroflow_expect_warnings","hydroflow_parser","hydroflow_syntax","hydroflow_syntax_noemit","monotonic_fn","morphism","rassert","rassert_eq","var_args","var_expr","var_type"],"mod":["compiled","props","scheduled","util"]};