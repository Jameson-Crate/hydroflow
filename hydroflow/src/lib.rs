#![feature(never_type)]
#![feature(type_alias_impl_trait)]
#![feature(impl_trait_in_assoc_type)]
#![allow(type_alias_bounds)]
#![allow(clippy::let_and_return)]
#![allow(clippy::iter_with_drain)]
#![allow(clippy::explicit_auto_deref)]

pub mod compiled;
pub mod lang;
pub mod props;
pub mod scheduled;
pub mod util;

pub use variadics::{self, var_args, var_expr, var_type};
pub use {
    bincode, bytes, futures, lattices, pusherator, rustc_hash, serde, serde_json, tokio,
    tokio_stream, tokio_util,
};

mod declarative_macro;
pub use declarative_macro::*;
#[cfg(feature = "hydroflow_datalog")]
pub use hydroflow_datalog::*;
#[cfg(feature = "hydroflow_macro")]
pub use hydroflow_macro::*;

#[cfg(doctest)]
mod booktest {
    macro_rules! booktest {
        ($i:ident, $( $t:tt )*) => {
            #[doc = include_str!(concat!("../../docs/docs/", $( stringify!($t), )* stringify!($i), ".md"))]
            mod $i {}
        };
    }

    booktest!(example_1_simplest, quickstart/);
    booktest!(example_2_simple, quickstart/);
    booktest!(example_3_stream, quickstart/);
    booktest!(example_4_neighbors, quickstart/);
    booktest!(example_5_reachability, quickstart/);
    booktest!(example_6_unreachability, quickstart/);
    booktest!(example_7_echo_server, quickstart/);
    booktest!(example_8_chat_server, quickstart/);

    booktest!(index, syntax/);
    booktest!(surface_embedding, syntax/);
    booktest!(surface_flows, syntax/);
    booktest!(surface_data, syntax/);

    mod surface_ops {
        hydroflow_macro::surface_booktest_operators!();
    }
}
