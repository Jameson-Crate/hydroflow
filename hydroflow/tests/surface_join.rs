use multiplatform_test::multiplatform_test;

use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};

#[multiplatform_test]
pub fn test_join_empty_lhs() {
    let (result_send, _) = hydroflow::util::unbounded_channel::<(u32, u32)>();

    let mut hf = hydroflow_syntax! {
        join_node = join::<'tick, 'tick>() -> for_each(|x: (u32, (u32, u32))| result_send.send(x.1).unwrap());
        null() -> [0] join_node;
        repeat_iter(1..) -> map(|x| (x, x)) -> [1] join_node;
    };
    assert_graphvis_snapshots!(hf);

    for tick in 0..10 {
        assert_eq!(tick, hf.current_tick());
        hf.run_tick();
    }
}

#[multiplatform_test]
pub fn test_join_empty_rhs() {
    let (result_send, _) = hydroflow::util::unbounded_channel::<(u32, u32)>();

    let mut hf = hydroflow_syntax! {
        join_node = join::<'tick, 'tick>() -> for_each(|x: (u32, (u32, u32))| result_send.send(x.1).unwrap());
        repeat_iter(1..) -> map(|x| (x, x)) -> [0] join_node;
        null() -> [1] join_node;
    };
    assert_graphvis_snapshots!(hf);

    for tick in 0..10 {
        assert_eq!(tick, hf.current_tick());
        hf.run_tick();
    }
}
