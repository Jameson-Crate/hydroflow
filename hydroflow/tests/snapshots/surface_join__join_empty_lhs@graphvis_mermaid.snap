---
source: hydroflow/tests/surface_join.rs
expression: hf.meta_graph().unwrap().to_mermaid()
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#02f,color:#fff,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff0,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    3v1[\"(3v1) <code>null()</code>"/]:::pullClass
    4v1[\"(4v1) <code>repeat_iter(1..)</code>"/]:::pullClass
    5v1[\"(5v1) <code>map(|x| (x, x))</code>"/]:::pullClass
    1v1[\"(1v1) <code>join::&lt;'tick, 'tick&gt;()</code>"/]:::pullClass
    2v1[/"(2v1) <code>for_each(|x: (u32, (u32, u32))| result_send.send(x.1).unwrap())</code>"\]:::pushClass
    3v1--0--->1v1
    4v1--->5v1
    5v1--1--->1v1
    1v1--->2v1
    subgraph sg_1v1_var_join_node ["var <tt>join_node</tt>"]
        1v1
        2v1
    end
end

